gslib/tests/test_cp.py

```diff
@@ -1002,6 +1002,42 @@ def _Check2():
 
     _Check2()
 
+  @SkipForXML('No compressed transport encoding support for the XML API.')
+  @SequentialAndParallelTransfer
+  def test_request_reason_header(self):
+    """Tests that x-goog-request-reason is set successfully on copy."""
+    bucket_uri = self.CreateBucket()
+    dst_uri = suri(bucket_uri, 'foo')
+    fpath = self._get_test_file('test.gif')
+
+    self.RunGsUtil(['-h', 'X-Goog-Request-Reason:b/this_is_reason', 'cp',
+                    fpath, dst_uri])
+    stdout = self.RunGsUtil(['ls', '-L', dst_uri], return_stdout=True)
+    self.assertRegex(stdout, r'X-Goog-Request-Reason:b/this_is_reason')
+    dst_uri2 = suri(bucket_uri, 'bar')
+    self.RunGsUtil(['cp', dst_uri, dst_uri2])
+    # Ensure metadata was preserved across copy.
+    stdout = self.RunGsUtil(['ls', '-L', dst_uri2], return_stdout=True)
+    self.assertRegex(stdout, r'X-Goog-Request-Reason:b/this_is_reason')
+
+  @SkipForXML('No compressed transport encoding support for the XML API.')
+  @SequentialAndParallelTransfer
+  def test_request_reason_set_by_environment_var(self):
+    """Tests that request reason can be set successfully using the environment
+    variable CLOUDSDK_CORE_REQUEST_REASON."""
+    os.environ['CLOUDSDK_CORE_REQUEST_REASON'] = 'b/this_is_env_reason'
+    bucket_uri = self.CreateBucket()
+    dst_uri = suri(bucket_uri, 'foo')
+    fpath = self._get_test_file('test.gif')
+    self.RunGsUtil(['cp', fpath, dst_uri])
+    stdout = self.RunGsUtil(['ls', '-L', dst_uri], return_stdout=True)
+    self.assertRegex(stdout, r'X-Goog-Request-Reason:b/this_is_env_reason')
+    dst_uri2 = suri(bucket_uri, 'bar')
+    self.RunGsUtil(['cp', dst_uri, dst_uri2])
```

Comment:
I don't think we need this second part if we're already testing it in `test_request_reason_header()`.

Reply:
Thanks for the suggestion. I change the second part to test when both the environment variable and http header is set, the request reason comes from the http header